fix: fix RpcDB connection so one can attach local Instance to remote DB for debugging purpose

kamilchodola opened the pull request:
Fixes RpcDB mode - with proper configuration it allows to debug remote DB (or at least use some remote data and debug RPC calls issues).

To make it work, make sure local node runs with flags:
 --Init.DiagnosticMode=RpcDb --Init.RpcDbUrl=JsonRPC URL (properly formatted with http:// as prefix)
 Also enabling of debug rpc endpoints is needed as in background few debug calls are made to fetch from remote DB
 
 Also for now it is quite slow (at least on my end) so worth applying --Init.ProcessingEnabled=false and disabling CL for processing on the desired node so will avoid messages about potential corruption of Db and will easily proceed to a enabled rpc node

## Types of changes

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [X] No

kamilchodola committed:
Add single block receipt miration functionality

kamilchodola committed:
Improve test and add missing param pass

kamilchodola committed:
add rpcDB improvements

kamilchodola committed:
Revert "Improve test and add missing param pass"

This reverts commit abd8d396691756e2b1c5b38abdee937be6da5ee8.

kamilchodola committed:
Revert "Add single block receipt miration functionality"

This reverts commit 1cf11ae524a6aa14236636c3b3deb69c9bbed937.

kamilchodola committed:
Adjust stateDb name

kamilchodola committed:
Adjustments

kamilchodola committed:
Merge branch 'master' into kch/rpc-db

LukaszRozmej commented on src/Nethermind/Nethermind.Db.Rpc/RpcDb.cs:
```diff
@@ -96,7 +97,9 @@ private byte[] GetThroughRpc(ReadOnlySpan<byte> key)
             if (response.Result is not null)
             {
-                value = Bytes.FromHexString((string)response.Result);
+                JsonElement jsonElement = (JsonElement)response.Result;
+                string rawHex = jsonElement.GetString();
+                value = Bytes.FromHexString(rawHex);
```
```suggestion
                value = Bytes.FromHexString(response.Result.ToString());
```

kamilchodola committed:
Update src/Nethermind/Nethermind.Db.Rpc/RpcDb.cs

Co-authored-by: Lukasz Rozmej <lukasz.rozmej@gmail.com>

kamilchodola commented on src/Nethermind/Nethermind.Db/FullPruning/FullPruningInnerDbFactory.cs:
```diff
@@ -59,7 +59,7 @@ private DbSettings GetRocksDbSettings(DbSettings originalSetting)
 
             // if first DB, then we will put it into main directory and not use indexed subdirectory
-            string dbName = firstDb ? originalSetting.DbName : originalSetting.DbName + _index;
+            string dbName = originalSetting.DbName;
```
But path still point to proper location - just DbName is different - without this change RPC was looking for a DB called "State0"

LukaszRozmej replied:
Should be fine 'Path' is more important and 'Name' is mostly for our convenience. You might want to change `FullPruningInnerDbFactoryTests.MatchSettings`

kamilchodola replied:
Applied — updated `FullPruningInnerDbFactoryTests.MatchSettings` for the new un-indexed DbName in 4c7ce6e. Since the factory now always uses `originalSetting.DbName`, the helper no longer appends the index to the expected name (`DbName + index` -> `DbName`); the path assertion still uses the index for the subdirectory. All 3 FullPruningInnerDbFactory tests pass locally.

kamilchodola committed:
refactor(db): address RpcDb review comments

Use string interpolation for the RPC hex payload and update
FullPruningInnerDbFactoryTests.MatchSettings for the un-indexed DbName.

kamilchodola commented on src/Nethermind/Nethermind.Db.Rpc/RpcDb.cs:
```diff
@@ -96,7 +97,7 @@ private byte[] GetThroughRpc(ReadOnlySpan<byte> key)
             if (response.Result is not null)
             {
-                value = Bytes.FromHexString((string)response.Result);
+                value = Bytes.FromHexString(response.Result.ToString());
```
Applied in 4c7ce6e — switched to `$"{response.Result}"`.
